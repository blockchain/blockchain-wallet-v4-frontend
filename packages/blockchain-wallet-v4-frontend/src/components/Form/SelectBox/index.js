import React from 'react'
import PropTypes from 'prop-types'
import styled from 'styled-components'

import { SelectInput } from 'blockchain-info-components'

const Container = styled.div`
  position: relative;
  width: 100%;
  height: auto;
  z-index: ${(props) => props.zIndex || 'initial'};
  background-color: ${(props) => props.theme.white};
`
const Error = styled.label`
  position: absolute;
  top: ${(props) => (props.errorBottom ? '48px' : '-20px')};
  right: 0;
  display: block;
  height: 15px;
  font-size: 12px;
  font-weight: 400;
  font-family: 'Inter', -apple-system, BlinkMacSystemFont, 'Segoe UI', Roboto, Oxygen, Ubuntu,
    Cantarell, 'Open Sans', 'Helvetica Neue', sans-serif;
  color: ${(props) => props.theme.error};
`

class SelectBox extends React.PureComponent {
  componentDidUpdate(prevProps) {
    if (this.props.meta.active && !prevProps.meta.active) {
      this.selectRef.focus()
    }
  }

  getSelectRef = (node) => {
    if (node) this.selectRef = node
  }

  onKeyPressed = (evt) => {
    const event = evt || window.event
    if (event.keyCode === 27) {
      event.stopPropagation()
      this.selectRef.blur()
    }
  }

  render() {
    const { className, errorBottom, hideErrors, input, meta, zIndex, ...rest } = this.props
    const { error, invalid, pristine, touched } = meta
    const errorState = touched && invalid ? 'invalid' : 'initial'

    return (
      <Container className={className} zIndex={zIndex} data-e2e='dropdownSelect'>
        <SelectInput
          {...input}
          {...meta}
          {...rest}
          onKeyDown={this.onKeyPressed}
          getRef={this.getSelectRef}
          errorState={errorState}
        />
        {(touched || !pristine) && error && !hideErrors && (
          <Error errorBottom={errorBottom}>{error}</Error>
        )}
      </Container>
    )
  }
}

SelectBox.propTypes = {
  elements: PropTypes.arrayOf(
    PropTypes.shape({
      group: PropTypes.string,
      items: PropTypes.arrayOf(
        PropTypes.shape({
          text: PropTypes.oneOfType([
            PropTypes.string.isRequired,
            PropTypes.number.isRequired,
            PropTypes.object.isRequired
          ]),
          value: PropTypes.oneOfType([
            PropTypes.string.isRequired,
            PropTypes.number.isRequired,
            PropTypes.object.isRequired
          ])
        })
      )
    })
  ).isRequired,
  input: PropTypes.shape({
    onBlur: PropTypes.func.isRequired,
    onChange: PropTypes.func.isRequired,
    onFocus: PropTypes.func.isRequired,
    value: PropTypes.oneOfType([PropTypes.string, PropTypes.number, PropTypes.object])
  }).isRequired,
  label: PropTypes.oneOfType([PropTypes.string, PropTypes.object]),
  opened: PropTypes.bool,
  searchEnabled: PropTypes.bool
}

export default SelectBox
