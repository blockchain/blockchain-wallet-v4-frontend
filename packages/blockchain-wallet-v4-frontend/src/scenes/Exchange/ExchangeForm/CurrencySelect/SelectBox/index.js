import React from 'react'
import styled from 'styled-components'
import { pathOr, prop, toLower } from 'ramda'

import { Icon } from 'blockchain-info-components'
import { SelectBox } from 'components/Form'

const ExchangeSelect = styled(SelectBox)`
  .bc__control {
    border: none;
    :hover {
      border: none;
    }
    :active {
      border: none;
    }
  }

  .bc__value-container {
    padding: 0;
  }
  .bc__option {
    padding: 8px 0;
    font-weight: 600;
  }

  .bc__single-value {
    color: ${props => props.theme.white};
  }
`

const DisplayWrapper = styled.div`
  border-radius: 3px;
  background-color: ${props => props.theme[props.coin]};
  min-height: 40px;
  display: flex;
  flex-direction: row;
  justify-content: flex-start;
  align-items: center;
  width: 100%;
  height: 100%;
  padding: 5px;
  box-sizing: border-box;
  cursor: pointer;
`
const ItemWrapper = styled.div`
  display: flex;
  flex-direction: row;
  justify-content: flex-start;
  align-items: center;
  width: 100%;
  height: 100%;
  padding: 5px;
  box-sizing: border-box;
  cursor: pointer;
  min-width: 0;
`
const Text = styled.span`
  position: relative;
  font-family: 'Montserrat', sans-serif;
  font-weight: 300;
  font-size: 14px;
  text-transform: ${props => (props.uppercase ? 'uppercase' : 'none')};
  font-style: normal;
  cursor: pointer;
  width: 100%;
  padding-left: 5px;
  white-space: nowrap;
  overflow: hidden;
  text-overflow: ellipsis;
`

const DisplayIcon = styled(Icon)`
  font-size: 28px;
  color: ${props => props.theme.white};
`

const selectItemIconColor = props =>
  props.isSelected ? props.theme.white : props.theme[props.coin]

const ItemIcon = styled(Icon)`
  font-size: 24px;
  color: ${selectItemIconColor} !important;
`

const getIconName = coin => `${toLower(coin)}`

const renderDisplay = (props, children) => {
  const coin = pathOr('', ['value', 'coin'], props)
  return (
    <DisplayWrapper className={props.className} coin={toLower(coin)}>
      {<DisplayIcon name={getIconName(coin)} />}
      <Text>{children}</Text>
    </DisplayWrapper>
  )
}

const renderItem = item => {
  const coin = pathOr('', ['value', 'coin'], item)
  const isSelected = prop('isSelected', item)
  return (
    <ItemWrapper>
      {
        <ItemIcon
          coin={toLower(coin)}
          isSelected={isSelected}
          name={getIconName(coin)}
        />
      }
      <Text>{item.text}</Text>
    </ItemWrapper>
  )
}

const SelectBoxExchange = props => (
  <ExchangeSelect
    {...props}
    searchEnabled={false}
    hideIndicator={true}
    templateDisplay={renderDisplay}
    templateItem={renderItem}
  />
)

export default SelectBoxExchange
