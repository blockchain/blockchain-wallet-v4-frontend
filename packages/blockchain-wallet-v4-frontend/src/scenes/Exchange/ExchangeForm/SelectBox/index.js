import React from 'react'
import styled from 'styled-components'
import { path } from 'ramda'

import { Icon } from 'blockchain-info-components'
import { SelectBox } from 'components/Form'

const ExchangeSelect = styled(SelectBox)`
  .bc__control {
    border-radius: 3px;
    border: none;
    background-color: ${props => props.theme[props.color]};
    :hover {
      border: none;
    }
  }

  .bc__value-container {
    padding: 0;
  }
  .bc__option {
    padding: 8px 0;
  }

  .bc__single-value {
    color: ${props => props.theme.white};
  }
`

const DisplayWrapper = styled.div`
  display: flex;
  flex-direction: row;
  justify-content: flex-start;
  align-items: center;
  width: 100%;
  height: 100%;
  padding: 5px;
  box-sizing: border-box;
  cursor: pointer;
`
const ItemWrapper = styled.div`
  display: flex;
  flex-direction: row;
  justify-content: flex-start;
  align-items: center;
  width: 100%;
  height: 100%;
  padding: 5px;
  box-sizing: border-box;
  cursor: pointer;
  min-width: 0;
`
const Text = styled.span`
  position: relative;
  font-family: 'Montserrat', sans-serif;
  font-weight: 300;
  font-size: 14px;
  text-transform: ${props => (props.uppercase ? 'uppercase' : 'none')};
  font-style: normal;
  cursor: pointer;
  width: 100%;
  padding-left: 5px;
  white-space: nowrap;
  overflow: hidden;
  text-overflow: ellipsis;
`

const DisplayIcon = styled(Icon)`
  font-size: 24px;
  color: ${props => props.theme.white};
`

const ItemIcon = styled(Icon)`
  font-size: 18px;
  color: ${props => props.theme[props.color]} !important;
`

const getIconName = coin => `exchange-${coin.toLowerCase()}`

const renderDisplay = (props, children) => {
  const coin = path(['value', 'coin'], props)
  return (
    <DisplayWrapper coin={coin} {...props}>
      {<DisplayIcon name={getIconName(coin)} />}
      <Text>{children}</Text>
    </DisplayWrapper>
  )
}

const renderItem = item => {
  const coin = path(['value', 'coin'], item)
  return (
    <ItemWrapper>
      {<ItemIcon color={coin.toLowerCase()} name={getIconName(coin)} />}
      <Text>{item.text}</Text>
    </ItemWrapper>
  )
}

const SelectBoxExchange = props => (
  <ExchangeSelect
    {...props}
    searchEnabled={false}
    color={path(['input', 'value', 'coin'], props).toLowerCase()}
    hideIndicator={true}
    templateDisplay={renderDisplay}
    templateItem={renderItem}
  />
)

export default SelectBoxExchange
