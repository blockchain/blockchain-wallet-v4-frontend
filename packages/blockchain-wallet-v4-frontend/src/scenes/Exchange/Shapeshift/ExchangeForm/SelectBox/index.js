import React from 'react'
import styled from 'styled-components'
import { path, prop } from 'ramda'

import { Icon } from 'blockchain-info-components'
import { SelectBox } from 'components/Form'

const DisplayWrapper = styled.div`
  display: flex;
  flex-direction: row;
  justify-content: flex-start;
  align-items: center;
  width: 100%;
  height: 100%;
  padding: 5px;
  box-sizing: border-box;
  cursor: pointer;
  min-width: 0;
`
const ItemWrapper = styled.div`
  display: flex;
  flex-direction: row;
  justify-content: flex-start;
  align-items: center;
  width: 100%;
  height: 100%;
  padding: 5px;
  box-sizing: border-box;
  cursor: pointer;
  min-width: 0;
`
const Text = styled.span`
  position: relative;
  font-family: 'Montserrat', sans-serif;
  font-weight: 300;
  font-size: 14px;
  text-transform: ${props => (props.uppercase ? 'uppercase' : 'none')};
  font-style: normal;
  color: ${props => props.theme['gray-5']};
  cursor: pointer;
  width: 100%;
  padding-left: 5px;
  white-space: nowrap;
  overflow: hidden;
  text-overflow: ellipsis;
`

const renderDisplay = item => (
  <DisplayWrapper>
    {prop('value', item) === 'BCH' && (
      <Icon name='bitcoin-cash' size='14px' weight={300} />
    )}
    {prop('value', item) === 'BTC' && (
      <Icon name='bitcoin' size='14px' weight={300} />
    )}
    {prop('value', item) === 'ETH' && (
      <Icon name='ethereum' size='14px' weight={300} />
    )}
    <Text>{item.text}</Text>
  </DisplayWrapper>
)

const renderItem = item => (
  <ItemWrapper>
    <Text>{item.text}</Text>
  </ItemWrapper>
)

const renderItemWithIcon = item => (
  <ItemWrapper>
    {path(['value', 'coin'], item) === 'BCH' && (
      <Icon name='bitcoin-cash' size='14px' weight={300} />
    )}
    {path(['value', 'coin'], item) === 'BTC' && (
      <Icon name='bitcoin' size='14px' weight={300} />
    )}
    {path(['value', 'coin'], item) === 'ETH' && (
      <Icon name='ethereum' size='14px' weight={300} />
    )}
    <Text>{item.text}</Text>
  </ItemWrapper>
)

const SelectBoxExchange = props => {
  return props.hasOneAccount ? (
    <SelectBox
      {...props}
      templateDisplay={renderDisplay}
      templateItem={renderItemWithIcon}
    />
  ) : (
    <SelectBox {...props} templateItem={renderItem} grouped />
  )
}

export default SelectBoxExchange
