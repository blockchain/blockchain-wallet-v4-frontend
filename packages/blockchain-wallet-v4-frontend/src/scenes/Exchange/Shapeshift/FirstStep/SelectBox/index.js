import React from 'react'
import styled from 'styled-components'
import { path, prop } from 'ramda'

import { Icon, Separator, Text } from 'blockchain-info-components'
import { SelectBox } from 'components/Form'

const DisplayWrapper = styled.div`
  display: flex;
  flex-direction: row;
  justify-content: flex-start;
  align-items: center;
  width: 100%;
  height: 100%;
  padding: 5px;
  box-sizing: border-box;
  cursor: pointer;
  text-overflow: ellipsis;

  & > * { margin-left: 5px; }
`
const HeaderWrapper = styled.div`
  display: flex;
  flex-direction: row;
  justify-content: flex-start;
  align-items: center;
  width: 100%;
  height: 100%;
  padding: 5px;
  box-sizing: border-box;
  cursor: not-allowed;
  text-overflow: ellipsis;

  & > * { margin-left: 5px; }
  & > :first-child { margin-right: 5px; }
`
const ItemWrapper = styled.div`
  display: flex;
  flex-direction: row;
  justify-content: flex-start;
  align-items: center;
  width: 100%;
  height: 100%;
  padding: 5px;
  box-sizing: border-box;
  cursor: pointer;
  text-overflow: ellipsis;

  & > * { margin-left: 5px; }
  &:hover {
    color: ${props => props.theme['gray-4']};
    background-color: ${props => props.theme['gray-1']};
  }
`
const renderDisplay = item => (
  <DisplayWrapper>
    {prop('value', item) === 'BCH' && <Icon name='bitcoin-cash' size='14px' weight={300} />}
    {prop('value', item) === 'BTC' && <Icon name='bitcoin' size='14px' weight={300} />}
    {prop('value', item) === 'ETH' && <Icon name='ethereum' size='14px' weight={300} />}
    <Text size='14px' weight={300}>{item.text}</Text>
  </DisplayWrapper>
)
const renderHeader = item => (
  <HeaderWrapper>
    {prop('text', item) === 'Bitcoin' && <Icon name='bitcoin-in-circle' size='14px' weight={300} />}
    {prop('text', item) === 'Bitcoin cash' && <Icon name='bitcoin-cash' size='14px' weight={300} />}
    {prop('text', item) === 'Ether' && <Icon name='ethereum-filled' size='14px' weight={300} />}
    <Separator align='right'>
      <Text size='14px' weight={300} uppercase>{item.text}</Text>
    </Separator>
  </HeaderWrapper>
)
const renderItem = item => (
  <ItemWrapper>
    <Text size='14px' weight={300}>{item.text}</Text>
  </ItemWrapper>
)

const renderItemWithIcon = item => (
  <ItemWrapper>
    {path(['value', 'coin'], item) === 'BCH' && <Icon name='bitcoin-cash' size='14px' weight={300} />}
    {path(['value', 'coin'], item) === 'BTC' && <Icon name='bitcoin' size='14px' weight={300} />}
    {path(['value', 'coin'], item) === 'ETH' && <Icon name='ethereum' size='14px' weight={300} />}
    <Text size='14px' weight={300}>{item.text}</Text>
  </ItemWrapper>
)

const SelectBoxExchange = props => props.hasOneAccount
  ? <SelectBox {...props} templateDisplay={renderDisplay} templateItem={renderItemWithIcon} />
  : <SelectBox {...props} templateDisplay={renderDisplay} templateHeader={renderHeader} templateItem={renderItem} />

export default SelectBoxExchange
