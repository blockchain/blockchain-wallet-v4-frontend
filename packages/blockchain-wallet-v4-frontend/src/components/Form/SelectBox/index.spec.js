import React from 'react'
import renderer from 'react-test-renderer'
import { mount } from 'enzyme'
import 'jest-styled-components'
import SelectBox from './index.js'
jest.mock('blockchain-info-components', () => ({
  SelectInput: ({ getRef, ...props }) => <input {...props} ref={getRef} />
}))

const fakeInput = {
  onBlur: jest.fn(),
  onChange: jest.fn(),
  onFocus: jest.fn(),
  value: ''
}

const fakeElements = [
  {
    group: 'fakeG',
    items: [
      { text: '1', value: 1 },
      { text: '2', value: 2 },
      { text: '3', value: 3 }
    ]
  }
]

describe('SelectBox', () => {
  it('renders correctly', () => {
    const props = { meta: {} }
    const component = renderer.create(
      <SelectBox {...props} input={fakeInput} elements={fakeElements} />
    )
    const tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('should focus inner ref on active props change', () => {
    const props = { meta: { active: false } }
    const wrapper = mount(
      <SelectBox {...props} input={fakeInput} elements={fakeElements} />
    )
    const focusSpy = jest.fn()
    wrapper.find('input').instance().focus = focusSpy

    wrapper.setProps({
      meta: { active: true }
    })
    expect(focusSpy).toHaveBeenCalledTimes(1)

    wrapper.setProps({
      meta: { active: false }
    })
    expect(focusSpy).toHaveBeenCalledTimes(1)

    wrapper.setProps({
      meta: { active: true }
    })
    expect(focusSpy).toHaveBeenCalledTimes(2)
  })
})
