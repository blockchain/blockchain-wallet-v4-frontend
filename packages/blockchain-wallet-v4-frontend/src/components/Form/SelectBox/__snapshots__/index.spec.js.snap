// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SelectBox renders correctly 1`] = `
.c0 {
  position: relative;
  width: 100%;
  height: auto;
}

<div
  className="c0"
>
  <select-input
    elements={
      Array [
        Object {
          "group": "fakeG",
          "items": Array [
            1,
            2,
            3,
          ],
        },
      ]
    }
    errorState="initial"
    onBlur={[MockFunction]}
    onChange={[MockFunction]}
    onFocus={[MockFunction]}
    value=""
  />
</div>
`;
